botor/botor.py

```diff
@@ -0,0 +1,211 @@
+            if fnmatch.fnmatch(api_params["Bucket"], mapping.get("source_bucket_pattern")):
+                if "Key" in api_params:
+                    if "source_key_pattern" in mapping:
+                        if not fnmatch.fnmatch(api_params["Key"], mapping.get("source_key_pattern")):
+                            continue
+                    if "mapped_prefix" in mapping:
+                        api_params["Key"] = mapping.get("mapped_prefix") + api_params["Key"]
+                if "mapped_bucket_name" in mapping:
+                    api_params["Bucket"] = mapping.get("mapped_bucket_name")
+
+                return map.get(profile), api_params
+    return map.get("default"), api_params
+
+
+def _route_list_params(kwargs, config, map):
+    if "Prefix" in kwargs:
+        client_to_call, result_args = _route_bucket_and_key(
+            api_params={"Bucket": kwargs.get("Bucket"), "Key": kwargs.get("Prefix")}, config=config, map=map)
+        kwargs["Prefix"] = result_args["Key"]
+    else:
+        client_to_call, result_args = _route_bucket_and_key(api_params=kwargs, config=config, map=map)
+    kwargs["Bucket"] = result_args["Bucket"]
+    return client_to_call, kwargs
+
+
+class PaginatorWrapper(object):
+    """Wrapper for a boto paginator.
+
+    Holds multiple paginators, one for each client, and dispatches calls to the appropriate
+    paginator according to Botor's mapping configuration
+    """
+
+    def __init__(self, mapping, config, operation_name):
+        """Init PaginatorWrapper.
+
+        Initialize paginator for each client.
+
+         :param dict mapping: The mapping between the profiles to the s3 clients
+         :param dict[dict] config: The configuration rules for the clients routing
+         :param str operation_name: The operation name of the paginator
+        """
+        self.mapping = mapping
+        self.config = config
+        self.paginators = dict()
+        for client in self.mapping:
+            self.paginators[client] = self.mapping[client].get_paginator(operation_name)
+
+    def paginate(self, **kwargs):
+        """iterate over the pages of the paginator API operation results.
+
+        accepts a PaginationConfig named argument that can be used to customize the pagination.
+        """
+        paginator_to_call, kwargs = _route_list_params(kwargs, self.config, self.paginators)
+        return getattr(paginator_to_call, "paginate")(**kwargs)
+
+
+class BotorBuilder(object):
+    """This class creates a botor client that wraps boto clients.
+
+    * Holds boto clients and routes the requests between them by bucket/prefix configuration.
+    * Create its methods on the fly according to boto3 client AWS methods.
+    * Holds special treatment for functions that operate on multiple buckets or keys
+    """
+
+    def __init__(self):
+        """Init BotorBuilder."""
+        self.default = None
+        self.mapping = None
+        self.config = None
+
+    def build(self, mapping, config):
```

Comment:
Any reason why I need to create the object and then call its `build` method?  Could we just have `__init__(self, mapping, config)` with this body?

Reply:
Yes, we must use the build method in order to generate the boto methods on the fly. ``python__init__(self, mapping, config)`` will not work

Reply:
Oh, I see!  So this is not really a class, just a function.

I'll refactor it (in a separate) to remove this class entirely.